cps/oauth_bb.py

```diff
@@ -22,328 +21,183 @@
+
+        log.debug("%s: %s %r", blueprint.name, request.path, request.args)
+        log.error("%s: %r", blueprint.name, error)
+
+        flash_text = None
+        if isinstance(error, ValueError):
+            if error.args[0] == "Cannot set OAuth token without an associated user":
+                # flash("%s authorization successful." % provider.name, category="info")
+                flash_text = "No associated local user found."
+
+        if not flash_text:
+            flash_text = (
+                u"OAuth error from {name}! "
+                u"{error} description={description} uri={uri}"
+            ).format(
+                name=provider.name,
+                error=error,
+                description=error_description,
+                uri=error_uri,
+            )
+        # ToDo: Translate
+        flash(flash_text, category="error")
+        provider.set_session_error(True)
+
+
+    def login(provider, success_view_name='web.index', failed_view_name='web.login'):
+        if provider.has_session_error():
+            log.debug("login %s: %s has error", provider.id, current_user)
+            # account authorized, but not linked?
+            return redirect(url_for(failed_view_name))
 
+        authorized = provider.is_authorized()
+        # log.debug("login %s: %s authorized=%s", provider.id, current_user, authorized)
+        if not authorized:
+            return redirect(url_for('%s.login' % provider.id))
 
-    # notify on OAuth provider error
-    @oauth_error.connect_via(oauthblueprints[0]['blueprint'])
-    def github_error(blueprint, error, error_description=None, error_uri=None):
-        msg = (
-            u"OAuth error from {name}! "
-            u"error={error} description={description} uri={uri}"
-        ).format(
-            name=blueprint.name,
-            error=error,
-            description=error_description,
-            uri=error_uri,
-        ) # ToDo: Translate
-        flash(msg, category="error")
-
-
-    @oauth.route('/github')
-    @oauth_required
-    def github_login():
-        if not github.authorized:
-            return redirect(url_for('github.login'))
-        account_info = github.get('/user')
-        if account_info.ok:
-            account_info_json = account_info.json()
-            return bind_oauth_or_register(oauthblueprints[0]['id'], account_info_json['id'], 'github.login')
-        flash(_(u"GitHub Oauth error, please retry later."), category="error")
-        return redirect(url_for('web.login'))
-
-
-    @oauth.route('/unlink/github', methods=["GET"])
+        # this should have already been cached by authorized_handler
+        account_id = provider.get_account_id()
+        if not account_id:
+            flash(_(u"%s Oauth error, please retry later." % provider.name), category="error")
+            return redirect(url_for(failed_view_name))
+
+        assert current_user.is_authenticated, "(%s) login successful, no user?" % provider.id
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846306)